firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,358 @@
+  private val session: ClientWebSocketSession?,
+  @Background private val backgroundDispatcher: CoroutineContext,
+  private var audioHelper: AudioHelper? = null
+) {
+
+  private val audioQueue = ConcurrentLinkedQueue<ByteArray>()
+  private val playBackQueue = ConcurrentLinkedQueue<ByteArray>()
+  private var startedReceiving = false
+  private var receiveChannel: Channel<Frame> = Channel()
+  private var isRecording: Boolean = false
+
+  private companion object {
+    val TAG = LiveSession::class.java.simpleName
+    val MIN_BUFFER_SIZE =
+      AudioTrack.getMinBufferSize(
+        24000,
+        AudioFormat.CHANNEL_OUT_MONO,
+        AudioFormat.ENCODING_PCM_16BIT
+      )
+  }
+
+  internal class ClientContentSetup(val turns: List<Content.Internal>, val turnComplete: Boolean) {
+    @Serializable
+    internal class Internal(@SerialName("client_content") val clientContent: ClientContent) {
+      @Serializable
+      internal data class ClientContent(
+        val turns: List<Content.Internal>,
+        @SerialName("turn_complete") val turnComplete: Boolean
+      )
+    }
+
+    fun toInternal() = Internal(Internal.ClientContent(turns, turnComplete))
+  }
+
+  @OptIn(ExperimentalSerializationApi::class)
+  internal class ToolResponseSetup(
+    val functionResponses: List<FunctionResponsePart.Internal.FunctionResponse>
+  ) {
+
+    @Serializable
+    internal data class Internal(val toolResponse: ToolResponse) {
+      @Serializable
+      internal data class ToolResponse(
+        val functionResponses: List<FunctionResponsePart.Internal.FunctionResponse>
+      )
+    }
+
+    fun toInternal() = Internal(Internal.ToolResponse(functionResponses))
+  }
+
+  internal class ServerContentSetup(val modelTurn: Content.Internal) {
+    @Serializable
+    internal class Internal(@SerialName("serverContent") val serverContent: ServerContent) {
+      @Serializable
+      internal data class ServerContent(@SerialName("modelTurn") val modelTurn: Content.Internal)
+    }
+
+    fun toInternal() = Internal(Internal.ServerContent(modelTurn))
+  }
+
+  internal class MediaStreamingSetup(val mediaChunks: List<MediaData.Internal>) {
+    @Serializable
+    internal class Internal(val realtimeInput: MediaChunks) {
+      @Serializable internal data class MediaChunks(val mediaChunks: List<MediaData.Internal>)
+    }
+    fun toInternal() = Internal(Internal.MediaChunks(mediaChunks))
+  }
+
+  internal data class ToolCallSetup(
+    val functionCalls: List<FunctionCallPart.Internal.FunctionCall>
+  ) {
+
+    @Serializable
+    internal class Internal(val toolCall: ToolCall) {
+
+      @Serializable
+      internal data class ToolCall(val functionCalls: List<FunctionCallPart.Internal.FunctionCall>)
+    }
+
+    fun toInternal(): Internal {
+      return Internal(Internal.ToolCall(functionCalls))
+    }
+  }
+
+  private fun fillRecordedAudioQueue() {
+    CoroutineScope(backgroundDispatcher).launch {
+      audioHelper!!.startRecording().collect {
```

Comment:
Should we propagate a `@RequiresPermission(android.Manifest.permission.RECORD_AUDIO)`?